### PR TITLE
[WIP] Integrate cuSpatial with gpuCI

### PR DESCRIPTION
**Pull request by dillon-cullinan**
Depends on: https://github.com/rapidsai/cuspatial/pull/39

- [x] GPU build.sh script
- [x] CPU build.sh scripts
- [x] Changelog check script
- [x] Style check script
- [x] Conda meta.yaml
- [x] Conda build.sh script
- [x] gpuci-scripts job updates https://github.com/rapidsai/gpuci-scripts/pull/60
- [x] Internal non-GPU testing

### CODE REVIEW

**Comment by dillon-cullinan** [comment]
At this point I have a strong outline of the basis for all the scripts. I would like @harrism to review some of the conda stuff. `cudf` has many header files that it tests in its `meta.yaml`, should I be including all of cuspatials header files in here similarly?

The dependencies are all the same as cudf right now since you currently require a cudf source build. Once you are able to build with a cudf conda package we can remove a lot of this. Reliance on the cudf source build also will heavily increase CI build times.

These changes are all untested for now. Will be focused on testing CPU builds tomorrow.

File: ci/cpu/prebuild.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+#Upload cuspatial once per CUDA
+if [[ "$CUDA" == "9.2" ]]; then
+    export UPLOAD_CUDF=1
```
Comment (raydouglass) [inline ci/cpu/prebuild.sh]: Rename these to `UPLOAD_CUSPATIAL`/`UPLOAD_LIBCUSPATIAL`

File: ci/gpu/build.sh [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+logger "Check environment..."
+env
+
+logger "Check GPU usage..."
+nvidia-smi
+
+logger "Activate conda env..."
+source activate gdf
+conda install "rmm=$MINOR_VERSION.*" "cudatoolkit=$CUDA_REL" \
```
Comment (raydouglass) [inline ci/gpu/build.sh]: I'm assuming the conda & pip installs are for cudf, not cuspatial? Can you document that so when a from-source build of cudf isn't necessary, these installs can be removed?

File: ci/gpu/build.sh [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+conda list
+
+################################################################################
+# BUILD - Build libnvstrings, nvstrings, libcudf, and cuDF from source
+#
+# cuSpatial currently requires a source build of cudf and not a conda package
+################################################################################
+
+logger "Clone cudf"
+git clone git@github.com:rapidsai/cudf.git ${CUDF_HOME}
```
Comment (raydouglass) [inline ci/gpu/build.sh]: You probably want specify which branch to clone: `git clone git@github.com:rapidsai/cudf.git -b branch-$MINOR_VERSION ${CUDF_HOME}`

File: conda/recipes/libcuspatial/build.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+# This assumes the script is executed from the root of the repo directory
+export CUDF_HOME="${WORKSPACE}/cudf"
+# show environment
+printenv
+# Cleanup local git
+git clean -xdf
+# build cudf with verbose output
+cd $CUDF_HOME
+./build.sh -v libnvstrings nvstrings libcudf cudf
```
Comment (raydouglass) [inline conda/recipes/libcuspatial/build.sh]: Does `libcuspatial` need `nvstrings` and `cudf` and not just the lib ones?

File: conda/recipes/cuspatial/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+requirements:
+  build:
+    - python
+    - cython >=0.29,<0.30
+    - setuptools
+    - numba >=0.45.1
+    - dlpack
+    - pyarrow 0.14.1.*
+    - libcudf {{ version }}
```
Comment (raydouglass) [inline conda/recipes/cuspatial/meta.yaml]: No dependency on `cudf`?

File: conda/recipes/libcuspatial/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    - PARALLEL_LEVEL
+  run_exports:
+    - {{ pin_subpackage("libcuspatial", max_pin="x.x") }}
+
+requirements:
+  build:
+    - cmake >=3.12.4
+  host:
+    - librmm {{ minor_version }}.*
+    - libnvstrings {{ minor_version }}.*
```
Comment (raydouglass) [inline conda/recipes/libcuspatial/meta.yaml]: No dependency on `libcudf`?

File: conda/recipes/libcuspatial/build.sh (hunk 4 above)
Comment (dillon-cullinan) [reply conda/recipes/libcuspatial/build.sh]: Deferring to @harrism

File: conda/recipes/cuspatial/meta.yaml (hunk 5 above)
Comment (dillon-cullinan) [reply conda/recipes/cuspatial/meta.yaml]: Since we are building cudf/libcudf from source as part of the conda build process, I'm assuming it should not be part of the dependency list.

File: conda/recipes/libcuspatial/meta.yaml (hunk 6 above)
Comment (dillon-cullinan) [reply conda/recipes/libcuspatial/meta.yaml]: See above comment.

File: build.sh [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+   -v           - verbose build mode
+   -g           - build for debug
+   -n           - no install step
+   -h           - print this text
+   default action (no args) is to build and install 'libcuspatial' then
+   'cuspatial' targets
+"
+LIBCUSPATIAL_BUILD_DIR=${REPODIR}/cpp/build
+CUSPATIAL_BUILD_DIR=${REPODIR}/python/cuspatial/build
+BUILD_DIRS="${LIBNVSTRINGS_BUILD_DIR} ${NVSTRINGS_BUILD_DIR} ${LIBCUDF_BUILD_DIR} ${CUDF_BUILD_DIR} ${DASK_CUDF_BUILD_DIR}"
```
Comment (rlratzel) [inline build.sh]: BUILD_DIRS looks like it needs to be updated

**Comment by rlratzel** [comment]
> I'm also unclear on how needing to build cudf from source might affect this. Will the configure and build of cudf from source happen automatically, or does there need to be something that explicitly calls cudf's build.sh?

I now see that the `gpu`, recipes, etc. scripts call cudf's build.sh, but it seems like the root build.sh should take care of that instead, to be consistent.

**Comment by dillon-cullinan** [comment]
> > I'm also unclear on how needing to build cudf from source might affect this. Will the configure and build of cudf from source happen automatically, or does there need to be something that explicitly calls cudf's build.sh?
> 
> I now see that the `gpu`, recipes, etc. scripts call cudf's build.sh, but it seems like the root build.sh should take care of that instead, to be consistent.

cuSpatial requiring a cuDF source build is only temporary. Once they are able to use a conda packages, all cudf build artifacts and source build references will be removed and replaced with the cudf conda package and added to the dependency list for conda.

**Comment by rlratzel** [comment]
> > > I'm also unclear on how needing to build cudf from source might affect this. Will the configure and build of cudf from source happen automatically, or does there need to be something that explicitly calls cudf's build.sh?
> > 
> > 
> > I now see that the `gpu`, recipes, etc. scripts call cudf's build.sh, but it seems like the root build.sh should take care of that instead, to be consistent.
> 
> cuSpatial requiring a cuDF source build is only temporary. Once they are able to use a conda packages, all cudf build artifacts and source build references will be removed and replaced with the cudf conda package and added to the dependency list for conda.

Yeah, that's why I was thinking you have only the root build.sh take care of that so you don't have to rip out all the other calls from the other scripts when it changes. Not a big deal at this point if leaving it as-is is easier for you.

File: conda/recipes/libcuspatial/build.sh (hunk 4 above)
Comment (harrism) [reply conda/recipes/libcuspatial/build.sh]: It needs cudf. It doesn't use nvstrings itself but cudf probably depends on it.

File: build.sh [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+# NOTE: ensure all dir changes are relative to the location of this
+# script, and that this script resides in the repo dir!
+REPODIR=$(cd $(dirname $0); pwd)
+
+VALIDARGS="clean libcuspatial cuspatial -v -g -n -h"
+HELP="$0 [clean] [libcuspatial] [cuspatial] [-v] [-g] [-n] [-h]
+   clean        - remove all existing build artifacts and configuration (start
+                  over)
+   libcuspatial - build the nvstrings C++ code only
```
Comment (harrism) [inline build.sh]:
```suggestion
   libcuspatial - build the cuSpatial C++ code only
```

File: build.sh [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+   cuspatial    - build the nvstrings Python package
```
Comment (harrism) [inline build.sh]:
```suggestion
   cuspatial    - build the cuSpatial Python package
```

File: build.sh [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    fi
+    done
+fi
+
+################################################################################
+# Configure, build, and install libcuspatial
+if (( ${NUMARGS} == 0 )) || hasArg libcuspatial; then
+
+    mkdir -p ${LIBCUSPATIAL_BUILD_DIR}
+    cd ${LIBCUDF_BUILD_DIR}
```
Comment (kkraus14) [inline build.sh]: Missed copy paste?

```suggestion
    cd ${LIBCUSPATIAL_BUILD_DIR}
```

File: build.sh [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+          -DCMAKE_BUILD_TYPE=${BUILD_TYPE} ..
+    if [[ ${INSTALL_TARGET} != "" ]]; then
+        make -j${PARALLEL_LEVEL} install_cuspatial VERBOSE=${VERBOSE}
+    else
+        make -j${PARALLEL_LEVEL} cuspatial VERBOSE=${VERBOSE}
+    fi
+fi
+
+# Build and install the cuspatial Python package
+if (( ${NUMARGS} == 0 )) || hasArg cudf; then
```
Comment (kkraus14) [inline build.sh]:
```suggestion
if (( ${NUMARGS} == 0 )) || hasArg cuspatial; then
```

File: ci/checks/changelog.sh [hunk 12]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+# Copyright (c) 2018, NVIDIA CORPORATION.
+#########################
+# cuDF CHANGELOG Tester #
```
Comment (kkraus14) [inline ci/checks/changelog.sh]:
```suggestion
# cuSpatial CHANGELOG Tester #
```

File: ci/checks/style.sh [hunk 13]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+# Copyright (c) 2018, NVIDIA CORPORATION.
+#####################
+# cuDF Style Tester #
```
Comment (kkraus14) [inline ci/checks/style.sh]:
```suggestion
# cuSpatial Style Tester #
```

File: ci/checks/style.sh [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+# Run black and get results/return code
+BLACK=`black --check python`
+BLACK_RETVAL=$?
+
+# Run flake8 and get results/return code
+FLAKE=`flake8 python`
+FLAKE_RETVAL=$?
+
+# Run flake8-cython and get results/return code
+FLAKE_CYTHON=`flake8 --config=python/cudf/.flake8.cython`
```
Comment (kkraus14) [inline ci/checks/style.sh]:
```suggestion
FLAKE_CYTHON=`flake8 --config=python/cuspatial/.flake8.cython`
```

File: ci/cpu/build.sh [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+# Logger function for build status output
+function logger() {
+  echo -e "\n>>>> $@\n"
+}
+
+# Set path and build parallel level
+export PATH=/conda/bin:/usr/local/cuda/bin:$PATH
+export PARALLEL_LEVEL=4
+export CUDF_HOME="${WORKSPACE}/cudf"
```
Comment (kkraus14) [inline ci/cpu/build.sh]:
```suggestion
export CUSPATIAL_HOME="${WORKSPACE}/cuspatial"
```

File: ci/cpu/libcuspatial/build_libcuspatial.sh [hunk 16]
```diff
@@ -0,0 +1,5 @@
+set -e
+
+echo "Building libcuspatial"
+conda build conda/recipes/libcuspatial --python=$PYTHON
```
Comment (kkraus14) [inline ci/cpu/libcuspatial/build_libcuspatial.sh]: Does this depend on a Python version? It shouldn't as far as I know:
```suggestion
conda build conda/recipes/libcuspatial
```

File: ci/cpu/prebuild.sh [hunk 17]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+#Upload cuspatial once per CUDA
```
Comment (kkraus14) [inline ci/cpu/prebuild.sh]: Only upload on CUDA 9.2 so it uploads on both Python 3.6 and 3.7
```suggestion
#Upload cuspatial once per PYTHON
```

File: ci/cpu/prebuild.sh [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+if [[ "$CUDA" == "9.2" ]]; then
+    export UPLOAD_CUSPATIAL=1
+else
+    export UPLOAD_CUSPATIAL=0
+fi
+
+#Upload libcuspatial once per PYTHON
```
Comment (kkraus14) [inline ci/cpu/prebuild.sh]: Only upload on Python 3.6 so it uploads on both CUDA 9.2 and CUDA 10.0
```suggestion
#Upload libcuspatial once per CUDA
```

File: ci/cpu/upload_anaconda.sh [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+  echo "Skipping upload"
+  return 0
+fi
+
+if [ -z "$MY_UPLOAD_KEY" ]; then
+    echo "No upload key"
+    return 0
+fi
+
+if [ "$UPLOAD_LIBCUDF" == "1" ]; then
```
Comment (kkraus14) [inline ci/cpu/upload_anaconda.sh]:
```suggestion
if [ "$UPLOAD_LIBCUSPATIAL" == "1" ]; then
```

File: conda/recipes/cuspatial/meta.yaml (hunk 5 above)
Comment (kkraus14) [reply conda/recipes/cuspatial/meta.yaml]: It will eventually, but not quite yet.

File: conda/recipes/libcuspatial/build.sh (hunk 4 above)
Comment (kkraus14) [reply conda/recipes/libcuspatial/build.sh]: Shouldn't it only need `libnvstrings` and `libcudf`?

File: ci/cpu/libcuspatial/build_libcuspatial.sh (hunk 16 above)
Comment (dillon-cullinan) [reply ci/cpu/libcuspatial/build_libcuspatial.sh]: I believe it actually does for now. libcuspatial builds libnvstrings, nvstrings, libcudf and cudf as part of its conda build.

File: conda/recipes/libcuspatial/build.sh (hunk 4 above)
Comment (harrism) [reply conda/recipes/libcuspatial/build.sh]: It needs cudf bindings in the Python API, I assume.

File: ci/cpu/libcuspatial/build_libcuspatial.sh (hunk 16 above)
Comment (kkraus14) [reply ci/cpu/libcuspatial/build_libcuspatial.sh]: With #39 it won't 😄

**Comment by dillon-cullinan** [comment]
I believe this is good to merge along with https://github.com/rapidsai/gpuci-scripts/pull/60.

GitHub webhook was added to the repo. CPU tests passed. Cannot test GPU builds as a GPU build is not available on dev Jenkins, so we can check those on another PR and if there are issues with it I can open another PR to fix those problems.

File: ci/cpu/upload_anaconda.sh [hunk 20]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+#
+# Adopted from https://github.com/tmcdonell/travis-scripts/blob/dfaac280ac2082cd6bcaba3217428347899f2975/update-accelerate-buildbot.sh
+
+set -e
+
+export LIBCUSPATIAL_FILE=`conda build conda/recipes/libcuspatial --python=$PYTHON --output`
```
Comment (kkraus14) [inline ci/cpu/upload_anaconda.sh]:
```suggestion
export LIBCUSPATIAL_FILE=`conda build conda/recipes/libcuspatial --output`
```

File: ci/cpu/upload_anaconda.sh [hunk 21]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+#
+# Adopted from https://github.com/tmcdonell/travis-scripts/blob/dfaac280ac2082cd6bcaba3217428347899f2975/update-accelerate-buildbot.sh
```
Comment (kkraus14) [inline ci/cpu/upload_anaconda.sh]: I don't really think our script is based off this anymore, remove?

File: ci/cpu/build.sh [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+python --version
+gcc --version
+g++ --version
+conda list
+
+# FIX Added to deal with Anancoda SSL verification issues during conda builds
+conda config --set ssl_verify False
+
+##########################################################################################
+# BUILD - Conda package builds (conda deps: libcupatial <- libcuspatial_cffi <- cuspatial)
```
Comment (kkraus14) [inline ci/cpu/build.sh]:
```suggestion
# BUILD - Conda package builds (conda deps: libcupatial <- cuspatial)
```

File: conda/recipes/libcuspatial/build.sh [hunk 23]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2018-2019, NVIDIA CORPORATION.
+
+# This assumes the script is executed from the root of the repo directory
+# show environment
+printenv
+# build cuspatial with verbose output
+cd $WORKSPACE
+echo "Fix permissions on build script"
+chmod ugo+x ./build.sh
```
Comment (kkraus14) [inline conda/recipes/libcuspatial/build.sh]: Why is this needed? Should the permissions be fixed on the file and then committed to the repo instead?

File: ci/gpu/build.sh [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+env
+
+logger "Check GPU usage..."
+nvidia-smi
+
+logger "Activate conda env..."
+source activate gdf
+
+# These installs are dependencies of cudf and most should be removed once we use a conda package
+# for cudf
```
Comment (kkraus14) [inline ci/gpu/build.sh]: Stale comment now?

File: ci/gpu/build.sh [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+conda install "rmm=$MINOR_VERSION.*" "cudf=$MINOR_VERSION" "cudatoolkit=$CUDA_REL"
```
Comment (kkraus14) [inline ci/gpu/build.sh]: Be consistent with minor version usage.

File: ci/gpu/build.sh [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+
+logger "Check versions..."
+python --version
+$CC --version
+$CXX --version
+conda list
+
+################################################################################
+# BUILD - Build libnvstrings, nvstrings, libcudf, and cuDF from source
```
Comment (kkraus14) [inline ci/gpu/build.sh]: Stale comment

File: ci/gpu/build.sh [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+function hasArg {
+    (( ${NUMARGS} != 0 )) && (echo " ${ARGS} " | grep -q " $1 ")
+}
+
+# Set path and build parallel level
+export PATH=/conda/bin:/usr/local/cuda/bin:$PATH
+export PARALLEL_LEVEL=4
+export CUDA_REL=${CUDA_VERSION%.*}
+export CUDF_HOME="${WORKSPACE}/cudf"
+export PYTHONPATH=${WORKSPACE}/python/cuspatial
```
Comment (kkraus14) [inline ci/gpu/build.sh]: We shouldn't do this anymore.
```suggestion
```

File: ci/gpu/build.sh (hunk 25 above)
Comment (kkraus14) [inline ci/gpu/build.sh]: We install rmm explicitly here, but don't in the conda recipes. Let's be consistent about it.
